proposed/activity-improvements.md

```diff
@@ -0,0 +1,326 @@
+
+- .NET Libraries have the `Activity` class which the object allows libraries and apps to publish tracing information. OT currently under development and not released yet. OT currently proposing the `Span` class which mostly used for the same purpose of Activity. OT Span supports more properties than Activity which allows more tracing scenarios. This proposal is offering the additions needed to be added to the Activity class which helps to address the same scenarios that Span does and possibly allow OT to fully get rid of Span to reduce the confusion which developers can run into to decide if need to use Activity or use Span. We are introducing the following types and properties which will be used inside the Activity class:
+    - ActivityContext which conforms to the w3c TraceContext specification. The context contains the trace Id, Span Id, Trace state and Trace flags.
+    - ActivityLink which can point to ActivityContexts inside a single Trace or across different Traces. Links can be used to represent batched operations where Activity was initiated by multiple initiating Activities each representing a single incoming item being processed in the batch.
+    - ActivityKind describes the relationship between the Activity, its parents, and its children in a Trace.
+    - Allow attaching any object to the Activity object through the `Set/GetCustomProperty` for the sake of the extendability without manually subclassing Activity or maintaining a dictionary mapping activity to the other objects. examples of that, OT implementation still depends on Activity and maintains a table mapping from Activity to Span. another example is some teams were working around the lack of missing Links and Contexts features and manually doing extra work to include these features. Although we are adding the Links and Contexts, we can get more future scenarios to extend Activity.
+
+- The current usage pattern of the Activity is not neat. here is an example of the current usage:
+```C#
+var activityListener = new DiagnosticSource("Azure.Core.Http");
+
+// Outer check to see if anyone is subscribed
+if (activityListener.IsEnabled())
+{
+    Activity activity = null;
+    // Check if anyone cares about activity
+    if (activityListener.IsEnabled("Azure.Core.Http.Request"))
+    {
+        activity = new Activity("Azure.Core.Http.Request");
+        activity.AddTag..
+        activityListener.StartActivity(activity); // this does string concat and allocates every time
+    }
+
+    ...
+
+    if (activity != null)
+    {
+        activityListener.StopActivity(activity);
+    }
+}
+```
+
+Note in the pattern, the code trying to avoid creating the Activity object if there are no listeners interested to listen to such Activity creation. Also, the code later has to check if the activity object is created and then stop it.
```

Comment:
> Note in the pattern, the code trying to avoid creating the Activity object if there are no listeners interested to listen to such Activity creation.

For perf.  In the below pattern, does that mean we're foregoing the perf benefits?

Reply:
The pattern below should be `Activity? activity = …`
We are not forgoing the benefits : )